Create Blog Post “make-app-to-phone-call-using-android-and-flutter” (#771)
Nexmo/deved-platform: content/blog/en/make-app-to-phone-call-using-flutter.md

Update body of the `_CallWidgetState` class:

```dart
class _CallWidgetState extends State<CallWidget> {
  SdkState _sdkState = SdkState.LOGGED_OUT;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            SizedBox(height: 64),
            _updateView()
          ],
        ),
      ),
    );
  }

  Widget _updateView() {
    if (_sdkState == SdkState.LOGGED_OUT) {
      return ElevatedButton(
          onPressed: () {  },
          child: Text("LOGIN AS ALICE")
      );
    }
  }
}
```

The initial state of the Flutter application is `SdkState.LOGGED_OUT`. 

Run the application you should see `Login Alice` button:

![](/content/blog/make-app-to-phone-call-using-flutter/loggedout.png)

### Login the user

> Button is disable so now add `onPressed` handler to the `ElevatedButton` to allow logging in:

Reviewer: Yes it's disabled - this is the default flutter behavior when you add the button without the handler

    }  else if (_sdkState == SdkState.WAIT) {
      return Center(
        child: CircularProgressIndicator(),
      );
    } else if (_sdkState == SdkState.LOGGED_IN) {
      return ElevatedButton(
          onPressed: () { _makeCall(); },
          child: Text("MAKE PHONE CALL")
      );
    }
  }
```

During `SdkState.WAIT` progress bar will be displayed. After succesfull login application will show `MAKE PHONE CALL` button.

> NOTE: While modyfying Android native code Flutter hot reload will not work. You have to stop the application and run it again.

![](/content/blog/make-app-to-phone-call-using-flutter/flutter-plugin-ui.png)

Run the app. Click `LOGIN AS ALICE` button. You should see `MAKE PHONE CALL` button (another state of the Flutter app based on the `SdkState` enum`):

![](/content/blog/make-app-to-phone-call-using-flutter/makeaphonecall.png)

### Make a call

To make a phone call open `main.dart` file and update body of `_makeCall` method:

```dart
Future<void> _makeCall() async {
    try {
      await requestPermissions();

      await platformMethodChannel
          .invokeMethod('makeCall');
    } on PlatformException catch (e) {
      print(e);
    }
  }
```

> The above method will communicate with Android so you have to update code in `MainActivity` class as well. Add `makeCall` clausule to `when` statement inside `addFlutterChannelListener` method:

Reviewer 2: Sorry @igorwojda , not sure what the word `clausule` is.  Did you mean `clause`?

Reviewer: Yes

```kotlin
private fun addFlutterChannelListener() {
        MethodChannel(flutterEngine?.dartExecutor?.binaryMessenger, "com.vonage").setMethodCallHandler { call, result ->

            when (call.method) {
                "loginUser" -> {
                    val token = requireNotNull(call.argument<String>("token"))
                    login(token)
                    result.success("")
                }
                "makeCall" -> {
                    makeCall()
                    result.success("")
                }
                else -> {
                    result.notImplemented()
                }
            }
        }
    }
```

> Now in the same file add `onGoingCall` property:

Reviewer 2: ```suggestion
Now in the same file add the `onGoingCall` property, which defines if and when a call is ongoing:
```

```kotlin
private var onGoingCall: NexmoCall? = null
```

> > NOTE: Currently Client SDK does not store ongoing call reference, so you have to store it in `MainActivity` class. You will use it later to end the call.

Reviewer 2: ```suggestion
> NOTE: Currently the Client SDK does not store ongoing call reference, so you have to store it in `MainActivity` class. You will use it later to end the call.
```

Now in the same file add `makeCall` method:

```kotlin
@SuppressLint("MissingPermission")
    private fun makeCall() {
        notifyFlutter(SdkState.WAIT)

        // Callee number is ignored because it is specified in NCCO config
        client.call("IGNORED_NUMBER", NexmoCallHandler.SERVER, object : NexmoRequestListener<NexmoCall> {
            override fun onSuccess(call: NexmoCall?) {
                onGoingCall = call
                notifyFlutter(SdkState.ON_CALL)
            }

            override fun onError(apiError: NexmoApiError) {
                notifyFlutter(SdkState.ERROR)
            }
        })
    }
```

> The above method sets the state of the Flutter app to `SdkState.WAIT` and waits for the Client SDK response (error or success). Now you need to add support for both states (`SdkState.ON_CALL` and `SdkState.ERROR`) inside `main.dart` file (Fluttter). Update body of the `_updateView` method:

Reviewer 2: ```suggestion
The above method sets the state of the Flutter app to `SdkState.WAIT` and waits for the Client SDK response (error or success). Now you need to add support for both states (`SdkState.ON_CALL` and `SdkState.ERROR`) inside `main.dart` file (Fluttter). Update body of the `_updateView` method to show the same as below:
```

```dart
Widget _updateView() {
    if (_sdkState == SdkState.LOGGED_OUT) {
      return ElevatedButton(
          onPressed: () { _loginUser(); },
          child: Text("LOGIN AS ALICE")
      );
    } else if (_sdkState == SdkState.WAIT) {
      return Center(
        child: CircularProgressIndicator(),
      );
    } else if (_sdkState == SdkState.LOGGED_IN) {
      return ElevatedButton(
          onPressed: () { _makeCall(); },
          child: Text("MAKE PHONE CALL")
      );
    } else if (_sdkState == SdkState.ON_CALL) {
      return ElevatedButton(
          onPressed: () { _endCall(); },
          child: Text("END CALL")
      );
    } else {
      return Center(
        child: Text("ERROR")
      );
    }
  }
```

> Each state change will result in UI modification. Before making a call the application needs specific permission.

Reviewer 2: ```suggestion
Each state change will result in UI modification. Before making a call the application needs specific permissions to use the microphone. In the next step, we're going to add the functionality in our project to request these permissions.
```

### Request permissions

The application needs to be able to access the microphone, so you have to request Android `android.permission.RECORD_AUDIO` permission (Flutter calls it `Permission.microphone`). 

First you need to add the [permission_handler](https://pub.dev/packages/permission_handler) package. Open `pubspec.yaml` file and add `permission_handler: ^6.0.1+1` dependency under `sdk: flutter`:

```yaml
dependencies:
  flutter:
    sdk: flutter

  permission_handler: ^5.1.0+2
```

> > NOTICE: Intention matters in `yaml` files, so make sure `permission_handler` is at the same indention level as the `flutter:` item.

Reviewer 2: ```suggestion
> NOTICE: Indentation matters in `yaml` files, so make sure `permission_handler` is at the same indentation level as the `flutter:` item.
```

Run the below command in the terminal to download the newly added Flutter package:

```cmd
flutter pub get
```

> Add package import at the top of the `main.dart` file:

Reviewer 2: ```suggestion
At the top of the `main.dart` file, you'll need to import the `permission_handler` package as shown in the example below:
```

```dart
import 'package:permission_handler/permission_handler.dart';
```

> Add this method inside `_CallWidgetState` class defined in the `main.dart` file to request permission:

Reviewer 2: To trigger the request for certain permissions, you'll need to add the `requestPermissions()` method within the `_CallWidgetState` class inside the `main.dart` file. So add this new method inside the class:

```dart
Future<void> requestPermissions() async {
    await [ Permission.microphone] .request();
  }
```

> Finally you need add two permissions (`uses-permission` tags) inside `app/src/main/AndroidManifest.xml` file, over the `application` tag:

Reviewer 2: ```suggestion
Finally, you need to add two permissions (`uses-permission` tags) inside `app/src/main/AndroidManifest.xml` file, above the `application` tag:
```

```xml
<uses-permission android:name="android.permission.INTERNET" />
<uses-permission android:name="android.permission.RECORD_AUDIO" />

<application
...
```

> NOTE: `android.permission.INTERNET` permission is granted implicitly by the Android, so you don't have to request it in Flutter explicitly.

> Run the app and click `MAKE PHONE CALL` to start a call. Permissions dialog will appear and after granting the permissions the Call will start.

Reviewer 2: ```suggestion
Run the app and click `MAKE PHONE CALL` to start a call. The permissions dialogue will appear and after granting the permissions the call will start.
```

> > Remainder: You define the phone number earlier in NCCO 

Reviewer 2: Other than the spelling mistake, could I ask what the relevance is to this statement? 

```suggestion
> Reminder: You defined the phone number earlier in NCCO 
```

> The state of the application will be updated to `SdkState.ON_CALL` and UI will be updated:

Reviewer 2: ```suggestion
The state of the application will be updated to `SdkState.ON_CALL` and the UI will be updated:
```

![](/content/blog/make-app-to-phone-call-using-flutter/oncall.png)

### End call

> Update body of the `_endCall` method inside 

Reviewer 2: Can we add a little bit to explain what we're doing in this section, why are we adding code to end the call? 
And with the sentence above `Update body of the...` inside where are we adding this method?

```dart
Future<void> _endCall() async {
    try {
      await platformMethodChannel.invokeMethod('endCall');
    } on PlatformException catch (e) {}
  }
```

> The above method will call communicate with Android so you have to update code in `MainActivity` class. Add `endCall` clausule to `when` statement inside `addFlutterChannelListener` method:

Reviewer 2: There's a suggestion for change here. And I'm not sure what `clausule` is. Is this `clause`?

```suggestion
The above method will communicate with Android so you have to update code in the `MainActivity` class. Add `endCall` clausule to `when` statement inside the `addFlutterChannelListener` method:
```

```kotlin
when (call.method) {
                "loginUser" -> {
                    val token = requireNotNull(call.argument<String>("token"))
                    login(token)
                    result.success("")
                }
                "makeCall" -> {
                    makeCall()
                    result.success("")
                }
                "endCall" -> {
                    endCall()
                    result.success("")
                }
                else -> {
                    result.notImplemented()
                }
            }
```

> Now in the same file add `endCall` method:

Reviewer 2: ```suggestion
Now in the same file add the `endCall` method:
```

```kotlin
private fun endCall() {
        onGoingCall?.hangup(object : NexmoRequestListener<NexmoCall> {
            override fun onSuccess(call: NexmoCall?) {
                onGoingCall = null
                notifyFlutter(SdkState.LOGGED_IN)
            }

            override fun onError(apiError: NexmoApiError) {
                notifyFlutter(SdkState.ERROR)
            }
        })
    }
```

> The above method sets the state of the Flutter app to `SdkState.WAIT` and waits for the Client SDK response (error or success). Both UI states are already supported in the Flutter application.

Reviewer 2: ```suggestion
The above method sets the state of the Flutter app to `SdkState.WAIT` and waits for the response from the Client SDK, which can be either error or success. Both UI states are already supported in the Flutter application.
```

Define `callEventListener` property at the top of the `MainActivity` class:

```kotlin
private val callEventListener = object : NexmoCallEventListener {
        override fun onMemberStatusUpdated(callMemberStatus: NexmoCallMemberStatus, callMember: NexmoCallMember) {
            if (callMemberStatus == NexmoCallMemberStatus.COMPLETED || callMemberStatus == NexmoCallMemberStatus.CANCELLED) {
                onGoingCall = null
            }
        }

        override fun onMuteChanged(mediaActionState: NexmoMediaActionState, callMember: NexmoCallMember) {}

        override fun onEarmuffChanged(mediaActionState: NexmoMediaActionState, callMember: NexmoCallMember) {}

        override fun onDTMF(dtmf: String, callMember: NexmoCallMember) {}
    }
```

The `onMemberStatusUpdated` callback informs you about call end.

To register above listener modify `onSuccess` callback inside `makeCall` method: 

```kotlin
onGoingCall = call
onGoingCall?.addCallEventListener(callEventListener)
```

Finally modify `endCall` method to unregister the `callEventListener` listener inside `onSuccess` callback:

```kotlin
onGoingCall?.removeCallEventListener(callEventListener)
onGoingCall = null
```

> Run the app and test if everything is working as expected.

Reviewer 2: ```suggestion
Run the app and if you've followed through this tutorial step by step, you'll be able to make a phone call from your mobile application to a physical phone number.
```
